Creates stubbed percentile aggregate functions

rkennedy-mode opened the pull request:
Includes PERCENTILE_1, PERCENTILE_5, PERCENTILE_25, PERCENTILE_75, PERCENTILE_95, and PERCENTILE_99 aggregate functions as well as a MEDIAN function, which is functionally equivalent to PERCENTILE_50. The underlying implementation is stubbed and simply returns the percentile value (e.g. PERCENTILE_25 returns 25, PERCENTILE_75 returns 75, etc).

A t-digest based implementation will come soon. This commit is to unblock teams choosing to begin integrating the new aggregate functions into their codebases.

Supports mode/issues#2732.

rkennedy-mode committed:
Creates stubbed percentile aggregate functions

Inludes PERCENTILE_1, PERCENTILE_5, PERCENTILE_25,
PERCENTILE_75, PERCENTILE_95, and PERCENTILE_99 aggregate
functions as well as a MEDIAN function, which is functionally
equivalent to PERCENTILE_50. The underlying implementation is
stubbed and simply returns the percentile value (e.g.
PERCENTILE_25 returns 25, PERCENTILE_75 returns 75, etc).

A t-digest based implementation will come soon. This commit
is to unblock teams choosing to begin integrating the new
aggregate functions into their codebases.

zacMode commented on src/frontend/org/voltdb/planner/PlanAssembler.java:
```diff
@@ -113,6 +115,15 @@ public ParsedResultAccumulator(boolean orderIsDeterministic,
         }
     }
 
+    private static final Set<ExpressionType> PERCENTILE_EXPRESSIONS = Collections.synchronizedSet(EnumSet.of(
```
(i'm probably missing something here).. why does this need to be a `synchronizedSet`?  Could `unmodifiableSet` work instead? i.e. doesn't seem like this is written to again, so you could get a similar effect with less overhead..

rkennedy-mode replied:
Good call. That's me blindly following the recommendations in the EnumSet documentation without considering how my use case differs.

Switched to `unmodifiableSet`.

zacMode reviewed (dismissed):
One small comment, but doesn't need to hold things up. That said, this was more of general review as i don't have a deep understanding of c++ or volt. So I won't be offended if you want to hold off for some more knowledgeable input :) .

rkennedy-mode committed:
Switch EnumSet from synchronized to read-only

It's meant to be used read-only anyway and the synchronization
was adding unnecessary overhead.

Nice catch, Zac.

mstockm commented on src/ee/executors/aggregateexecutor.cpp:
```diff
@@ -602,6 +659,36 @@ inline Agg* getAggInstance(Pool& memoryPool, ExpressionType agg_type, bool isDis
+    case EXPRESSION_TYPE_AGGREGATE_MEDIAN:
+        return new (memoryPool) PercentileAgg(50);
+    case EXPRESSION_TYPE_AGGREGATE_TDIGEST_TO_MEDIAN:
+        return new (memoryPool) TDigestToPercentileAgg(50);
```
Nice that you found a way to not have to do the actual implementation n times and can parameterize it at least at one level :P

rkennedy-mode replied:
Indeed. If VoltDB would allow all of the necessary changes to live a single `.cpp` or `.h` file then I could use the preprocessor to write a template and then stamp it out N times (once for each supported percentile). Alas, the changes are **all over the place** and Java doesn't have a preprocessor.

mstockm reviewed (approved):
Seems to match existing code to me! 🙃

rkennedy-mode committed:
Merge branch 'mode-release-8.x' into add-stubbed-percentile-functions
